### PR TITLE
[WIP] Uniform handling of metaclasses

### LINKED ISSUE #2305: Support iteration on enums
Iterating on enums makes mypy (0.4.5) raise errors even though it's OK:

``` python
import enum


class Colors(enum.Enum):
    RED = 0
    BLUE = 1


print([color.name for color in Colors])
```

```
> python3 color.py 
['RED', 'BLUE']
```

```
> mypy color.py 
color.py:9: error: Iterable expected
color.py:9: error: "Colors" has no attribute "__iter__"
```

(follow-up issue created as suggested in #529)

### PR DESCRIPTION
**Pull request by elazarg**
In order to fix #2305, I think we need a better handling of metaclasses: store the type of the metaclass in TypeInfo, and use it when possible, instead of `builtins.type`.

I assume metaclasses inherit from type; I don't yet follow [the documentation](https://docs.python.org/3/reference/datamodel.html#determining-the-appropriate-metaclass).

This PR requires changes to typeshed.

### CODE REVIEW

**Comment by elazarg** [comment]
For some reason (circularity?) the tests do not recognize `abc.ABCMeta`, but do recognize `from abc import ABCMeta`

**Comment by elazarg** [comment]
How do I test changes to typeshed?

**Comment by JukkaL** [comment]
Typeshed has a travis build that does some minimal consistency checks. I think that there's a script you can also run locally, but I haven't tried that. You can also run the mypy test suite locally against your typeshed changes. `pythoneval.test` in the mypy repo has tests that run against the full stubs, so if your stub changes are tightly coupled to changes to mypy internals, you can add some tests there. Note that they are much slower than most other tests, so we generally avoid adding tests there unless we really want to test interactions with real typeshed stubs.

More generally, you can write some dummy code that uses your changed stubs and manually try to verify that type checking and running a program seem compatible.

If your typeshed and mypy changes happen to depend on each other for a clean build, you'll probably want to coordinate them with a mypy core dev, as one of the builds may have to be red at least momentarily until both changes have been merged.

**Comment by elazarg** [comment]
Thanks

בתאריך יום ו׳, 28 באוק' 2016, 16:17, מאת Jukka Lehtosalo ‏<
notifications@github.com>:

**Comment by JukkaL** [comment]
(I didn't do a detailed review, just mentioned that it may better to split this PR.)

File: test-data/unit/pythoneval-enum.test [hunk 1]
```diff
@@ -118,3 +118,17 @@ takes_int(SomeExtIntEnum.x)
 def takes_some_ext_int_enum(s: SomeExtIntEnum):
     pass
 takes_some_ext_int_enum(SomeExtIntEnum.x)
+
+[case testEnumIter]
```
Comment (elazarg) [inline test-data/unit/pythoneval-enum.test]: There seem to be other tests relying on typeshed
```
FAILURE  #3 check file runtests.py

typeshed/stdlib/3/types.pyi:111: note: In module imported here,
typeshed/stdlib/3/sys.pyi:10: note: ... from here,
runtests.py:9: note: ... from here:
typeshed/stdlib/3/_importlib_modulespec.pyi:36: error: Name 'abc.ABCMeta' is not defined
typeshed/stdlib/3/_importlib_modulespec.pyi:36: error: Invalid metaclass 'abc.ABCMeta'
```

**Comment by gvanrossum** [comment]
> I think that there's a script you can also run locally, but I haven't tried that.

Here's the way I do it.
- I have a virtualenv with Python 3 enabled
- My typeshed repo is a subdirectory of my mypy repo
- cd into the typeshed repo
- git pull or edit or whatever you want to test
- PYTHONPATH=.. python3 tests/mypy_test.py

This runs 5 sets of tests:
```
$ PYTHONPATH=.. python3 tests/mypy_test.py
running mypy --python-version 3.5 --strict-optional # with 342 files
running mypy --python-version 3.4 --strict-optional # with 342 files
running mypy --python-version 3.3 --strict-optional # with 327 files
running mypy --python-version 3.2 --strict-optional # with 326 files
running mypy --python-version 2.7 --strict-optional # with 380 files
```
You can limit it to a single Python version using `-p2` or `-p3.5`.

**Comment by elazarg** [comment]
I think I will wait for syncing typeshed and reopen another one.

**Pull request closed without merging**